src/expr/macro/src/lib.rs

```diff
@@ -431,6 +431,8 @@ struct UserFunctionAttr {
     name: String,
     /// The last argument type is `&mut dyn Write`.
     write: bool,
+    /// The last argument type is `retract: bool`.
+    retract: bool,
```

Comment:
What about making this a const-generic? 👀

Reply:
It doesn't help for better performance and makes macro more complex. 🥵

  | retract: bool | Const RETRACT: bool | Cmp
-- | -- | -- | --
count_agg_vis_rate_0_05 | 146 | 147 | 1%
count_agg_vis_rate_0_25 | 443 | 430 | -3%
count_agg_vis_rate_0_5 | 858 | 844 | -2%
count_agg_vis_rate_0_75 | 1,187 | 1,192 | 0%
count_agg_without_vis | 950 | 1,083 | 14%
max_agg_vis_rate_0_05 | 156 | 164 | 5%
max_agg_vis_rate_0_25 | 490 | 513 | 5%
max_agg_vis_rate_0_5 | 926 | 982 | 6%
max_agg_vis_rate_0_75 | 1,306 | 1,385 | 6%
max_agg_without_vis | 1,080 | 954 | -12%
min_agg_vis_rate_0_05 | 156 | 164 | 5%
min_agg_vis_rate_0_25 | 471 | 498 | 6%
min_agg_vis_rate_0_5 | 926 | 956 | 3%
min_agg_vis_rate_0_75 | 1,305 | 1,398 | 7%
min_agg_without_vis | 1,081 | 952 | -12%
sum_agg_vis_rate_0_05 | 244 | 242 | -1%
sum_agg_vis_rate_0_25 | 858 | 855 | -0%
sum_agg_vis_rate_0_5 | 1,671 | 1,641 | -2%
sum_agg_vis_rate_0_75 | 2,598 | 2,447 | -6%
sum_agg_without_vis | 2,559 | 2,573 | 1%